### PR TITLE
backup, external_storage, cloud: make write functions asynchronous 

### LINKED ISSUE #11158: external_storage, cloud: make `put` an asynchronous function, and make backup asynchronous
## Development Task
The `write` interface of `ExternalStorage` is weird:

https://github.com/tikv/tikv/blob/98b53e0c04dc7d6a25553d6559e203609f6b7e94/components/external_storage/src/lib.rs#L58-L63

We accept an `AsyncRead`, but returning synchronously, in fact, all implementor of this interface involves `block_on_external_io` which create a temporary Tokio runtime, for executing a sole future. (How horrible... :( )

This also blocks https://github.com/tikv/tikv/issues/11112.

### PR DESCRIPTION
**Pull request by YuJuncen**
### What problem does this PR solve?

Issue Number: close #11158 

Problem Summary: See the issue.

### What is changed and how it works?

What's Changed:
1. Made `cloud::blob::BlobStorage::put` asynchronous.
2. Made `external_storage::ExternalStorage::write` asynchronous.
3. Made some wrapper for `Box<dyn AsyncRead + Unpin + Send>`, for make rustc happy (see the comment).
4. Replace the thread pool in the `backup::Endpoint` with `Tokio::Runtime`, and make backup async.
5. Refactor the `backup::softlimit::SoftLimit`, make it use a internal `Semaphore` to restrict concurrency.

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by YuJuncen** [comment]
fatal: index-pack failed 🤔
/build

**Comment by YuJuncen** [comment]
server::lock_manager::test_detect_deadlock_when_merge_region
/run-tests 

Still not merged

File: components/external_storage/src/lib.rs [hunk 1]
```diff
@@ -44,25 +45,27 @@ pub fn record_storage_create(start: Instant, storage: &dyn ExternalStorage) {
         .observe(start.saturating_elapsed().as_secs_f64());
 }
 
+/// UnpinReader is a simple wrapper for AsyncRead + Unpin + Send.
+/// This wrapper would remove the lifetime at the argument of the generted async function
+/// in order to make rustc happy. (And reduce the length of signture of write.)
+/// see https://github.com/rust-lang/rust/issues/63033
+pub struct UnpinReader(pub Box<dyn AsyncRead + Unpin + Send>);
```
Comment (kennytm) [inline components/external_storage/src/lib.rs:52]:
```suggestion
pub type UnpinReader = Box<dyn AsyncRead + Unpin + Send + 'static>;
```

?

File: components/cloud/src/blob.rs [hunk 2]
```diff
@@ -1,27 +1,49 @@
...
     fn name(&self) -> &'static str;
     fn url(&self) -> io::Result<url::Url>;
 }
 
+/// PutResource is a simple wrapper for put.
+/// It is identity to [external_storage::UnpinReader],
+/// only for decoupling external_storage and cloud package.
+///
+/// See the documentation of [external_storage::UnpinReader] for why those wrappers exists.
+pub struct PutResource(pub Box<dyn AsyncRead + Send + Unpin>);
```
Comment (kennytm) [inline components/cloud/src/blob.rs:21]: (ditto)

File: components/external_storage/src/lib.rs (hunk 1 above)
Comment (YuJuncen) [reply components/external_storage/src/lib.rs:52]: I'm afraid that we must use the new type pattern but type alias for resolving the lifetime deriving problem... Even through let me try it...🤔

File: components/backup/src/endpoint.rs [hunk 3]
```diff
@@ -716,13 +746,8 @@ impl<E: Engine, R: RegionInfoProvider + Clone + 'static> Endpoint<E, R> {
...
-        self.pool.borrow_mut().spawn(move || {
-            tikv_alloc::add_thread_memory_accessor();
+        self.pool.borrow_mut().spawn(async move {
             let _with_io_type = WithIOType::new(IOType::Export);
-            defer!({
-                tikv_alloc::remove_thread_memory_accessor();
-            });
 
             // Check if we can open external storage.
             let backend = match create_storage(&request.backend, config) {
```
Comment (3pointer) [inline components/backup/src/endpoint.rs:753]: this line can move outside.

**Review by 3pointer — APPROVED** [review]
Rest LGTM

File: components/backup/src/endpoint.rs (hunk 3 above)
Comment (YuJuncen) [reply components/backup/src/endpoint.rs:753]: Yes, but maybe let's done it in another PR?

File: components/backup/src/softlimit.rs [hunk 4]
```diff
@@ -1,97 +1,72 @@
...
+    async fn take_tokens(&self, n: usize) -> Result<()> {
+        self.0.semaphore.acquire_many(n as _).await?.forget();
         Ok(())
     }
 
-    fn grant_tokens(&self, n: usize) -> Result<()> {
-        for _ in 0..n {
-            self.sx.send(())?;
-        }
+    async fn grant_tokens(&self, n: usize) -> Result<()> {
```
Comment (kennytm) [inline components/backup/src/softlimit.rs:39]:
```suggestion
    async fn grant_tokens(&self, n: usize) {
```

this function never errors now.

File: components/backup/src/softlimit.rs [hunk 5]
```diff
@@ -1,97 +1,72 @@
...
+        self.take_tokens(n).await?;
         Ok(())
     }
 
     /// Grows the tasks can be executed concurrently by n
     #[cfg(test)]
-    pub fn grow(&self, n: usize) -> Result<()> {
-        self.cap.fetch_add(n, Ordering::SeqCst);
-        self.grant_tokens(n)?;
+    pub async fn grow(&self, n: usize) -> Result<()> {
```
Comment (kennytm) [inline components/backup/src/softlimit.rs:55]: ditto

File: components/backup/src/softlimit.rs [hunk 6]
```diff
@@ -236,53 +228,35 @@ mod softlimit_test {
...
-        should_finish_in(d, name, move || {
+    async fn should_satisfy_in(
+        d: Duration,
+        name: &str,
+        mut f: impl FnMut() -> bool + Send + 'static,
+    ) {
+        should_finish_in(d, name, async move {
             while !f() {
-                thread::sleep(d.div_f32(50.0));
+                time::sleep(d.div_f32(50.0)).await;
```
Comment (kennytm) [inline components/backup/src/softlimit.rs:253]:
```suggestion
                time::sleep(d / 50).await;
```

File: components/backup/src/writer.rs [hunk 7]
```diff
@@ -23,6 +25,31 @@ use txn_types::KvPair;
 use crate::metrics::*;
 use crate::{backup_file_name, Error, Result};
 
+#[derive(Debug, Clone, Copy)]
+/// CfNameWrap wraps the CfName type.
+/// For removing the 'static lifetime bound in the async function,
+/// which doesn't compile due to 'captures lifetime that does not appear in bounds' :(.
+/// FIXME: remove this.
```
Comment (kennytm) [inline components/backup/src/writer.rs:32]: add https://github.com/rust-lang/rust/issues/63033 to the comment for tracking.

**Review by kennytm — COMMENTED** [review]
rest LGTM

**Review by kennytm — APPROVED** [review]

**Pull request merged**